# Dataset Profiling

## Objective

Identify the different experiment types available in the NASA Battery Dataset.

The dataset contains three experiment categories:

- Charge
- Discharge
- Impedance

In [1]:
from pathlib import Path
import sys

PROJECT_ROOT = Path.cwd().parent

if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))


In [2]:
import pandas as pd

In [3]:
from src.analysis.dataset_analyzer import (
    analyze_battery,
    count_rows,
)

In [4]:
files = analyze_battery("B0005")

Battery : B0005
Total CSV Files : 616


## Number of Files

In [5]:
len(files)

616

## Dataset Size Analysis

In [6]:
rows = count_rows(files)

print(f"Minimum rows : {min(rows)}")
print(f"Maximum rows : {max(rows)}")
print(f"Average rows : {sum(rows)/len(rows):.2f}")

Minimum rows : 5
Maximum rows : 3900
Average rows : 981.82


## Column Format Identification

In [7]:
column_formats = {}

for file in files:

    df = pd.read_csv(file)

    columns = tuple(df.columns)

    if columns not in column_formats:
        column_formats[columns] = []

    column_formats[columns].append(file.name)

In [8]:
summary = []

for i, (columns, files) in enumerate(column_formats.items(), start=1):

    summary.append({
        "Format": i,
        "Number_of_Files": len(files),
        "Example_File": files[0],
        "Columns": ", ".join(columns)
    })

summary_df = pd.DataFrame(summary)

summary_df

,Format,Number_of_Files,Example_File,Columns
0,1,168,05122.csv,"Voltage_measured, Current_measured, Temperatur..."
1,2,170,05123.csv,"Voltage_measured, Current_measured, Temperatur..."
2,3,278,05161.csv,"Sense_current, Battery_current, Current_ratio,..."


## Experiment Type Distribution

In [9]:
for i, (columns, files) in enumerate(column_formats.items(), start=1):

    print("="*70)

    print(f"FORMAT {i}")

    print("="*70)

    print()

    print("Columns:")

    for c in columns:
        print("-", c)

    print()

    print("Number of files :", len(files))

    print("Example file :", files[0])

    print()

FORMAT 1

Columns:
- Voltage_measured
- Current_measured
- Temperature_measured
- Current_load
- Voltage_load
- Time

Number of files : 168
Example file : 05122.csv

FORMAT 2

Columns:
- Voltage_measured
- Current_measured
- Temperature_measured
- Current_charge
- Voltage_charge
- Time

Number of files : 170
Example file : 05123.csv

FORMAT 3

Columns:
- Sense_current
- Battery_current
- Current_ratio
- Battery_impedance
- Rectified_Impedance

Number of files : 278
Example file : 05161.csv



## Observations

Three experiment formats were identified.

### Format 1

Discharge Experiment

168 CSV files

### Format 2

Charge Experiment

170 CSV files

### Format 3

Impedance Experiment

278 CSV files

These observations will be used during preprocessing to automatically classify every CSV before feature engineering.